src/internal/collection/postgres_collection.go

```diff
@@ -675,14 +675,15 @@ func (c *postgresReadOnlyCollection) WatchByIndexF(index *Index, indexVal string
 
 type postgresReadWriteCollection struct {
 	*postgresCollection
-	tx *pachsql.Tx
+	ctx context.Context
```

Comment:
[Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it](https://pkg.go.dev/context#pkg-overview).

Some [notable](https://github.com/golang/go/issues/14660#issuecomment-200145301) [folks](https://github.com/golang/go/issues/14660#issue-138695431) suggested otherwise earlier on, but [this official post](https://go.dev/blog/context-and-structs) reflects subsequent experience.

What you want to do here is to thread the context through as the first argument to each call. So, e.g.:

```go
func (a *apiServer) someHandler(ctx context.Context, subject string, groups []string) error {
	return dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
		// ctx might be cancelled when WithTx is done
		groups := a.groups.ReadWrite(sqlTx)

		// Add user to new groups
		for group := range groups {
			if err := groups.Upsert(ctx, group, &membersProto, func(ctx context.Context) error {
				// ctx might be cancelled when the Upsert is done
				return nil
			}); err != nil {
				return errors.EnsureStack(err)
			}
		}
		return nil
	})
}
```

Written this way, func arguments to e.g. Upsert don’t have to be closures to use the enclosing context — they can just receive it as an argument.

I know that we already store a context in `postgresReadOnlyCollection`, but we shouldn’t repeat that mistake.